### PR TITLE
Improvements to Random Teleport

### LINKED ISSUE #1893: Random Spawn
Hi, so there one thing i feel essentialsx spawn needs. for servers trying to run in an anarchy setting its not uncommon for player to try to spawn trap people and for that reason essentials spawn needs to have a random feature. like a setting in the config when set to true would randomize a players spawn point at spawn whenever they died or did /spawn.  right now im using a command block with /spreadplayers command but thats not always reliable if the pressure plate that powers it is spammed. so please please make this a feature. thankyou for the good work.

### LINKED ISSUE #3469: RTP Multi World Support
### Feature request

**Feature description**
Add a section in the config to define what world a player will be randomly be teleported to from the world they are in. Example:

Spawn: Wilderness
Wilderness: Wilderness

If a player does /rtp in spawn, they will be randomly teleported in the Wilderness world. If they do /rtp in the Wilderness, they will be randomly teleported in the Wilderness world. Additionally, the addition of supporting multiple worlds that can be randomly teleported in.

**How the feature is useful**

Many servers have a spawn on a separate world. Many players run /wild or /rtp on their first command and get confused when they aren't getting RTP'd because it is disabled in the spawn world. This would solve that issue by randomly teleporting them into the desired world. Additionally, multi world support allows for server that have multiple worlds to be randomly teleported within.

### LINKED ISSUE #3479: Run TPR from console/command block
### Feature request

**Feature description**
Allow the console to run the TPR command passing the name of the player as an argument.

**How the feature is useful**
On the spawn of my wilderness world I have a button to get teleported randomly. Since the command can only be used by players the command block will not work

### PR DESCRIPTION
**Pull request by pop4959**
Supersedes #3538.
Closes #1893, closes #3469, closes #3479.

Various improvements for Essentials' random teleportation feature.

Should be complete as-is, other than potential code clean-up and testing before merging.

Miscellaneous points of interest:
- #easteregg `pre-cache` configuration removed (it was hardly in use anyways, and doesn't make any difference after first tp + with multiple locations the number of locations to compute could get very expensive).
- EssentialsUpgrade to convert old single location random teleport and retain prior behavior.
- Random respawn with `random-respawn` and `random-respawn-location` configuration options.
- Random teleport signs by adding `randomteleport` to the `enabledSigns` list in the config, writing `[RandomTeleport]` in the first line on the sign, and optionally the name of the location on the second line.
- Configuration option `default-location` for default location (where to teleport when typing /tpr with no arguments) that defaults to the current world.
- Configuration option `per-location-permission` for per location permissions (`essentials.tpr.location.<name>`), functioning similarly to how per warp permissions work.
- `/settpr` now requires the location name as an argument to set center/ranges for.
- Can teleport other players with the `essentials.tpr.others` and the command `/tpr <name> <player>`.
- Console can now execute `/tpr` with the name and player both as required arguments.
- Fixed an issue where on older versions of the game the default center location would be a block too high.
- Fixed an issue where events modifying random teleportation parameters would cause incorrect caching.
- Fixed the issue of random locations below y=0 being considered invalid (world height can be lower now in 1.17).

### CODE REVIEW

**Comment by JRoy** [comment]
@pop4959 can you resolve conflicts for this whenever you have time

**Comment by pop4959** [comment]
Updated, would like to see a few more people test this before merging though. Grab a build from [actions](https://github.com/EssentialsX/Essentials/actions/runs/1567843927) if you want to try it, and just be sure to backup tpr.yml so you can easily switch back to regular Essentials builds.

**Comment by Leahcimkrob** [comment]
You can enter a value after it for tpr, but you can always get by in the farmworld.

**Comment by pop4959** [comment]
> You can enter a value after it for tpr, but you can always get by in the farmworld.

Not quite sure what you mean here, can you elaborate?

**Comment by Leahcimkrob** [comment]
https://github.com/EssentialsX/Essentials/discussions/4679
I posted that in the discussions. So I got your invitation to test your build. With the / tpr I still got into the farming world, even if I had specified a value such as Nether.

**Comment by pop4959** [comment]
> #4679 I posted that in the discussions. So I got your invitation to test your build. With the / tpr I still got into the farming world, even if I had specified a value such as Nether.

Can you provide a list of the worlds you have on your server and your tpr.yml file

**Comment by Leahcimkrob** [comment]
Ok i got it.

After a few teleoports with / tpr, I suddenly landed in the Void in the Nether. Can that still be fixed?

**Review by Envel-Nikita-Gutsenkov — DISMISSED** [review]

**Review by Envel-Nikita-Gutsenkov — DISMISSED** [review]

**Comment by Chris6ix** [comment]
Hey, just wondering if there are any updates on this?
Just wanted to open a new issue, but saw that the Feature I want was already suggested (and apparently also been worked on) (#3469)
Would be awesome having this :)

Edit: Sadly can't help test because the build artifact is expired :/

**Comment by pop4959** [comment]
@Chris6ix Updated the build for you :)

This work is for the most part done, probably just a few minor changes here and there, but otherwise just needs to be tested thoroughly before the 2.20 milestone.

**Comment by Chris6ix** [comment]
Alright I downloaded it, and I'm going to test it when I got some spare time (Have Exams rn :/ )
Thanks for updating the build :)

Is there anything specific that needs to be tested or something?
I can imagine that the basics have already been tested?

Btw, is there an approximate release date for 2.20?

**Comment by pop4959** [comment]
> Alright I downloaded it, and I'm going to test it when I got some spare time (Have Exams rn :/ ) Thanks for updating the build :)
> 
> Is there anything specific that needs to be tested or something? I can imagine that the basics have already been tested?
> 
> Btw, is there an approximate release date for 2.20?

No problem. Nothing specific, just making sure multi-world teleports, /tpr, /settpr commands, configuration, etc all appear and behave as expected and are intuitive to use, not buggy, etc. The basics have been tested so ideally try it out on a server and see if you find any issues using it.

There isn't an ETA For 2.20 currently but you can track progress in the [milestone](https://github.com/EssentialsX/Essentials/milestone/14) here on Github.

**Comment by Chris6ix** [comment]
> No problem. Nothing specific, just making sure multi-world teleports, /tpr, /settpr commands, configuration, etc all appear and behave as expected and are intuitive to use, not buggy, etc. The basics have been tested so ideally try it out on a server and see if you find any issues using it.
> 
> There isn't an ETA For 2.20 currently but you can track progress in the [milestone](https://github.com/EssentialsX/Essentials/milestone/14) here on Github.

Okay thank you!
I will test these things when i got some spare time and get back to you with the results ;)

**Comment by Chris6ix** [comment]
Hey @pop4959 i found some free time and tested the multi world rtp feature :)
I loaded the 4 worlds from my Server to a completely fresh paper `1.18.2 #350` Test Server with only EssX `2.20.0-dev+6-2448e2a` (your build) and Multiverse-Core `4.3.2-SNAPSHOT-b870`
I also generated a completely new EssX folder and left everything to default.

I teleported myself to each of the world's spawns using `mvtp <WorldName>` and then just did `/settpr <WorldName> center`
This worked perfectly for all the worlds.

I then teleported myself to each world and made about 20 random teleports in each world, each time checking with [Xaeros World Map](https://www.curseforge.com/minecraft/mc-mods/xaeros-world-map) that the teleports were really random and that it was not picking a location more than once or something.
Everything worked perfectly fine besides the rtps in the nether...there were 2 issues:
Out of the 20 times, I got teleported 2 times into a wall
Out of the 20 times, I got teleported 3 times on top of the nether roof

So I think the nether part needs to be adjusted a bit with a max height just under the nether roof for example and a fix for the teleporting into walls. The rest looks fine to me :)

Thanks again for working on this great feature!
If you need any more testing or something, just ping me here.

Cheers

**Comment by pop4959** [comment]
@Chris6ix Hey, thanks for doing such a thorough test! Glad to hear everything's working for you. 😄 

Not sure about the teleporting into walls/roof, I don't believe any changes have been made with regards to that. It should be using Essentials safe teleport feature (as all other teleport commands do), so unless you've turned that off random teleports should behave correctly.

I sometimes forget to keep this branch up-to-date, but feel free to continue testing it. Just ping me if you need any changes merged in from the current mainstream dev builds.

**Comment by Chris6ix** [comment]
Hey @pop4959 thanks for the quick reply!
As I explained in my message from before, I generated a completely new EssX folder (so also a fresh config) and didn't even open it (the only file I checked was `tpr.yml` to check if all the worlds show up correctly). 
So all settings were on default, which for the teleport safety seems to be this:
image
So the teleport safety was enabled...and it still teleported me into walls / on top of the nether roof...
Or is that because I was in creative? (Because of explanation of `force-safe-teleport-location`)
If that is so, shouldn't the default be `true` to also prevent creative mode players from getting teleported into walls?

**Comment by pop4959** [comment]
Yes it is probably because you were in creative. The teleport is "safe" in creative mode since you do not take damage. Teleportation safety primarily concerns whether or not the player takes damage teleporting to the location.

You can test in survival or try setting that option to true if that is your preference.

**Comment by Chris6ix** [comment]
Mhm okay I have `force-safe-teleport-location` now set to `true` and I'm not getting teleported into walls anymore in creative, but it still teleports me on top of the nether roof (Probably because its "safe" because you don't take damage...)
Is there something you can do to not make the nether roof count as "safe" / Or add a max tp height just for the nether dimension just under the nether roof, which seems to be `Y:121` ?

Edit: This could be a config option if the nether roof should be treated as safe or not

**Comment by pop4959** [comment]
Okay, that's something we can look into then. The intent is definitely not to teleport on top of the roof. Probably either the location being selected too high, or teleportation safety choosing a spot above the roof (although I believe it's already supposed to do this).

**Comment by goweiwen** [comment]
Want to update that I've been using this PR on my server for more than a few months and it has been working great.

File: Essentials/src/main/java/com/earth2me/essentials/EssentialsUpgrade.java [hunk 1]
```diff
@@ -153,6 +153,10 @@ public static void uuidFileConvert(final IEssentials ess, final Boolean ignoreUF
         ess.getLogger().info("To rerun the conversion type /essentials uuidconvert");
     }
 
+    public void updateRandomTeleport() {
+        ess.getRandomTeleport().updateConfig();
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/EssentialsUpgrade.java]: A couple problems with this upgrade;
* This should only be ran once (by checking/setting the `doneFile` with a unique upgrade key)
* The logic for this upgrade should be contained in the `EssentialsUpgrade` class and ideally not call any method in `RandomTeleport` as to not create an accidental regression down the road.

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandsettpr.java [hunk 2]
```diff
@@ -17,22 +19,21 @@ public Commandsettpr() {
...
-        if (args.length == 0) {
+        if (args.length < 2) {
             throw new NotEnoughArgumentsException();
         }
-
         final RandomTeleport randomTeleport = ess.getRandomTeleport();
-        randomTeleport.getCachedLocations().clear();
-        if ("center".equalsIgnoreCase(args[0])) {
-            randomTeleport.setCenter(user.getLocation());
+        randomTeleport.reloadConfig();
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandsettpr.java]: This invokes an I/O operation. Is there a reason why it's ran let alone before any arguments are even validated?

File: Essentials/src/main/java/net/ess3/api/events/UserRandomTeleportEvent.java [hunk 3]
```diff
@@ -14,11 +14,19 @@ public class UserRandomTeleportEvent extends Event implements Cancellable {
...
     private boolean cancelled = false;
+    private boolean modified = false;
 
+    public UserRandomTeleportEvent(final IUser user, final String name, final Location center, final double minRange, final double maxRange) {
+        this(user, center, minRange, maxRange);
+        this.name = name;
+    }
+
+    @Deprecated
     public UserRandomTeleportEvent(final IUser user, final Location center, final double minRange, final double maxRange) {
```
Comment (JRoy) [inline Essentials/src/main/java/net/ess3/api/events/UserRandomTeleportEvent.java]: Event constructors aren't API, you can remove this if you're not using it

**Review by JRoy — CHANGES_REQUESTED** [review]
few things otherwise good!

**Comment by pop4959** [comment]
TODO
- jroys comments above
- pre-validate random locations in chunks
- refactor / clean up some code

**Review by JRoy — APPROVED** [review]

**Comment by JRoy** [comment]
This has beat the adventure pr for longest time from open to merge...

**Comment by PetyXbron** [comment]
🚀🚀🚀

**Comment by DavidDoesCode** [comment]
Im trying this out and have the follow config: (I tried many commands)

I'm finding that when I random tp - one of the coordinates are ALWAYS 1000 or -1000, which the other one is "random" 

It's spawning me on the "border" 

`default-location: '{mcverse_vanilla}'
excluded-biomes:
- cold_ocean
- deep_cold_ocean
- deep_frozen_ocean
- deep_lukewarm_ocean
- deep_ocean
- deep_warm_ocean
- frozen_ocean
- frozen_river
- lukewarm_ocean
- ocean
- river
- warm_ocean
locations:
  mcverse_vanilla:
    center:
      world: a58ce000-9699-4d7b-b91f-4644ce8955dc
      world-name: mcverse_vanilla
      x: 0.5
      y: 72.0
      z: 0.5
      yaw: -142.0070343017578
      pitch: 14.700003623962402
    min-range: 100.0
    max-range: 1000.0
  '{mcverse_vanilla}':
    center:
      world: bbfcef01-71a8-4676-8119-4cdad005180d
      world-name: world
      x: 0.0
      y: 63.0
      z: 0.0
      yaw: 0.0
      pitch: 0.0
    min-range: -1000.0
    max-range: 1000.0
`

**Comment by Evidentsinger14** [comment]
PR comments are not for support. if you need help with something, please use the support channels in [MOSS](https://discord.gg/h8CnPSw)

**Comment by pop4959** [comment]
@DavidDoesCode You should not have a negative range in your config. That is not supported.

**Comment by DavidDoesCode** [comment]
Maybe some better comments describing what the values should be.

But anyways, I'm pretty sure that the random respawn is not respecting beds:
`respawn-at-home-bed: true`

(I did try minrange with 0 but /essentials reload did not seem to work, restarting the server did)

And I didn't test but I have :
`respawn-at-anchor: true`

**Comment by pop4959** [comment]
@DavidDoesCode If the commands allow you to enter a negative radius, please open a bug report. Same goes for any other issues or concerns you have.

If you have any other questions please use the support channels in MOSS as Evident has mentioned.

**Pull request merged**